code/server/sv_http.c

```diff
@@ -0,0 +1,1066 @@
+		mem = (uint8_t *)malloc(client->requestLength);
+		memcpy(mem, client->request, client->requestLength);
+		request.proto.buf = mem;
+		request.proto.bufSize = client->requestLength;
+		if (!HTTP_ParseRequest(&request, mem, client->requestLength)) {
+			Com_DPrintf("Failed to parse request\n");
+			HTTP_AssembleError(client, HTTP_STATUS_InternalServerError);
+			continue;
+		}
+
+		FD_CLR(clientSocket, &httpReadFDSet);
+		FD_CLR(clientSocket, &readFDsOut);
+
+		memset(&response, 0, sizeof(response));
+		response.proto.contentLength = -1;
+		response.freeBody = qtrue;
+		response.status = HTTP_STATUS_Ok;
+		if (!HTTP_Route(client, &request, &response)) {
+			HTTP_AssembleError(client, HTTP_STATUS_NotFound);
+			continue;
+		}
+		if (client->responseLength == 0) {
+			HTTP_AssembleResponse(client, &response);
+			if (response.freeBody) {
+				if (response.file) {
+					FS_FreeFile((void*)response.body);
+				} else {
+					free((char *)response.body);
+				}
+			}
+		}
+		for (i = 0; i < HTTP_MAX_HEADERS; ++i) {
+			if (!response.proto.headers[i].freeValue) {
+				continue;
+			}
+			free(response.proto.headers[i].value);
+		}
+	}
+}
+
+static qboolean HTTP_ResponseSetPk3File(httpResponse_t *response, const char *filepath) {
+	long length;
+	fileHandle_t fileHandle;
+
+	length = FS_SV_FOpenFileRead(filepath, &fileHandle);
+	if (length <= 0) {
+		Com_Printf("Failed to load %s\n", filepath);
+		return qfalse;
+	}
+	response->body = (char*)malloc(length);
+	if (FS_Read((void*)response->body, (int)length, fileHandle) != (int)length) {
+		Com_Printf("Failed to read %s\n", filepath);
+		free((void*)response->body);
+		response->body = NULL;
+		return qfalse;
+	}
+
+	HTTP_SetHeader(&response->proto, HTTP_HEADER_CONTENT_TYPE, HTTP_MIMETYPE_APPLICATION_OCTET_STREAM, qfalse);
+	response->file = qfalse; // use normal free
+	response->freeBody = qtrue;
+	response->bodySize = length;
+	return qtrue;
+}
+
+static qboolean HTTP_ResponseSetFile(httpResponse_t *response, const char *filepath) {
+	const char *extension = COM_GetExtension(filepath);
+	long length;
+	length = FS_ReadFile(filepath, (void**)&response->body);
+	if (length <= 0) {
+		Com_Printf("Failed to load %s\n", filepath);
+		return qfalse;
+	}
+	if (!Q_stricmp(extension, "jpg") ||!Q_stricmp(extension, "jpeg")) {
+		HTTP_SetHeader(&response->proto, HTTP_HEADER_CONTENT_TYPE, HTTP_MIMETYPE_IMAGE_JPEG, qfalse);
+	} else if (!Q_stricmp(extension, "png")) {
+		HTTP_SetHeader(&response->proto, HTTP_HEADER_CONTENT_TYPE, HTTP_MIMETYPE_IMAGE_PNG, qfalse);
+	} else {
+		HTTP_SetHeader(&response->proto, HTTP_HEADER_CONTENT_TYPE, HTTP_MIMETYPE_APPLICATION_OCTET_STREAM, qfalse);
+	}
+	HTTP_SetHeader(&response->proto, HTTP_HEADER_CACHE_CONTROL, "public, max-age=31536000", qfalse);
+	response->file = qtrue;
+	response->freeBody = qtrue;
+	response->bodySize = length;
+	return qtrue;
+}
+
+static void HTTP_BodyHeader(char *buf, int size) {
+	// TODO: allow to customize stylesheet
```

Comment:
As much as I would like the whole header to be an external 'template' file that can customized, I don't think it's really needed unless there is a lot of people requesting it.